src/hal/src/acceleration_structure.rs

```diff
@@ -0,0 +1,224 @@
+    // - blas must be triangles or aabbs, tlas must be instances
+    pub geometries: &'a [AccelerationStructureGeometry<'a, B>],
+    // Both APIs support "array" and "array of pointers", presumably to allow for cheap reuse of the geometry descriptors.
+    // pgeometries: &'a [&'a GeometryDesc<'a, B>],
+    pub scratch: Option<(&'a B::Buffer, Offset)>,
+}
+
+bitflags! {
+    /// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkGeometryFlagBitsKHR.html
+    pub struct AccelerationStructureGeometryFlags: u32 {
+        /// indicates that this geometry does not invoke the any-hit shaders even if present in a hit group.
+        const OPAQUE = 0x1;
+        /// indicates that the implementation must only call the any-hit shader a single time for each primitive in this geometry. If this bit is absent an implementation may invoke the any-hit shader more than once for this geometry.
+        const NO_DUPLICATE_ANY_HIT_INVOCATION = 0x2;
+    }
+}
+
+#[derive(Debug)]
+pub struct AccelerationStructureGeometry<'a, B: Backend> {
+    pub flags: AccelerationStructureGeometryFlags,
+    pub geometry: Geometry<'a, B>,
+}
+
+#[derive(Debug)]
+pub enum Geometry<'a, B: Backend> {
+    Triangles(GeometryTriangles<'a, B>),
+
+    /// TODO bikeshed capitalization of AABBs.
+    Aabbs(GeometryAabbs<'a, B>),
+
+    Instances(GeometryInstances<'a, B>),
+}
+
+// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccelerationStructureGeometryTrianglesDataKHR.html
+// for docs: Memory Safety (https://nvpro-samples.github.io/vk_raytracing_tutorial_KHR/#accelerationstructure)
+#[derive(Debug)]
+pub struct GeometryTriangles<'a, B: Backend> {
+    /// Both APIs require support for at least the following:
+    /// - `(R32_G32, Float)`
+    /// - `(R32_G32_B32, Float)`
+    /// - `(R16_G16, Float)`
+    /// - `(R16_G16_B16_A16, Float)`
+    /// - `(R16_G16, Inorm)`
+    /// - `(R16_G16_B16_A16, Inorm)`
+    /// VK could support more by querying `VK_FORMAT_FEATURE_ACCELERATION_STRUCTURE_VERTEX_BUFFER_BIT_KHR`, DX12 is not queryable? Note [the DX12 ray tracing spec](https://microsoft.github.io/DirectX-Specs/d3d/Raytracing.html#d3d12_raytracing_geometry_triangles_desc) says it supports more than [the Win32 docs](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_raytracing_geometry_triangles_desc).
+    pub vertex_format: Format,
+    pub vertex_buffer: &'a B::Buffer,
+    pub vertex_buffer_offset: Offset,
+    pub vertex_buffer_stride: Offset,
+
+    /// aka "vertex count"?
+    pub max_vertex: Offset,
+
+    // index format must be DXGI_FORMAT_R32_UINT, DXGI_FORMAT_R16_UINT
+    // can also be VK_INDEX_TYPE_NONE_KHR/DXGI_FORMAT_UNKNOWN
+    pub index_buffer: Option<(&'a B::Buffer, Offset, IndexType)>,
+
+    /// 3x4 matrix
+    /// TODO enum for cpu repr?
```

Comment:
only Metal has this different? We can work around that by doing a copy, I think (either a transfer, or in compute)

Reply:
I believe this `transform` field is only read for presence during the "get sizes" phase and only actually read during the "build" phase. We should be in a command list at that point, so it's feasible to emulate the vulkan behavior like you say.

FYI, I think the table looks something like this:

| | Vulkan | DX12 | Metal |
| - | - | - | - |
| CPU repr | Y |  |  |
| GPU repr | Y | Y | Y* |

VK: [`VkDeviceOrHostAddressConstKHR transformData`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccelerationStructureGeometryTrianglesDataKHR.html)
DX: [`D3D12_GPU_VIRTUAL_ADDRESS Tansform3x4`](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_raytracing_geometry_triangles_desc)
MTL: *only on the "instance" type: [`transformBuffer: MTLBuffer?`](https://developer.apple.com/documentation/metalperformanceshaders/mpsinstanceaccelerationstructure)

Note in the vulkan link that it allows for all pointers in the `AccelerationStructureGeometryDesc` (i.e. `VkAccelerationStructureBuildGeometryInfoKHR`) hierarchy to be `VkDeviceOrHostAddressConstKHR`.

Do you think this is worth doing now or could I follow-up after a DX impl (I suspect this feature will be a bit of work there)?